src/txn/txn.c

```diff
@@ -1155,6 +1155,98 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
     return (ret);
 }
 
+/*
+ * __txn_commit_timestamps_verbose --
+ *     Print warning messages when encountering surprise timestamp usage.
+ */
+static inline int
+__txn_commit_timestamps_verbose(WT_SESSION_IMPL *session)
+{
+#ifndef HAVE_DIAGNOSTIC
+    WT_UNUSED(session);
+    return (0);
+#else
+#define WT_COMMIT_TS_VERB_PREFIX "Commit timestamp unexpected usage: "
+    WT_TXN *txn;
+    WT_TXN_OP *op;
+    WT_UPDATE *upd;
+    wt_timestamp_t op_ts, prev_op_durable_ts;
+    bool txn_has_ts;
+    char ts_string[2][WT_TS_INT_STRING_SIZE];
+    u_int i;
+
+    txn = session->txn;
+    txn_has_ts = F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) || F_ISSET(txn, WT_TXN_HAS_TS_DURABLE);
+
+    /* Nothing to do if no tables this transaction touched were configured for verbose logging. */
+    if (!F_ISSET(txn, WT_TXN_VERB_TS_WRITE))
+        return (0);
+
+    /* Don't check timestamp usage while a transaction is prepared. */
+    if (F_ISSET(txn, WT_TXN_PREPARE))
```

Comment:
Ideally this checking would be done while (or after) resolving prepared transactions, otherwise we need to be more careful about what we can say regards durable/commit.